crates/walrus-service/src/node.rs

```diff
@@ -2928,29 +3156,80 @@ impl ServiceState for StorageNodeInner {
         blob_id: BlobId,
         sliver_pair_index: SliverPairIndex,
         sliver: Sliver,
+        intent: UploadIntent,
     ) -> Result<bool, StoreSliverError> {
         self.check_index(sliver_pair_index)?;
+        let (metadata, persisted) = self
+            .resolve_metadata_for_sliver(&blob_id, intent.is_pending())
```

Comment:
Suggestion: instead of return error directly when no metadata is returned, we can also do a little bit of wait here. In this way, we can also parallelize metadata and sliver uploads.